### DISCUSSION TITLE
Ignore headers for S3 connection for other backends?

### DISCUSSION THREAD

**Discussion by gera2ld** [pocketbase/pocketbase#5917] [Q&A]
Hi there, thanks for this great project.

Recently I've encountered the same issue as #2231 , but with a different backend: SeaweedFS. I got SignatureDoesNotMatch error with Pocketbase while everything works with rclone.

I noticed that the previous fix in `ignoreSigningHeaders` only works when the endpoint contains `storage.googleapis.com`:

https://github.com/pocketbase/pocketbase/blob/b83f924c7d35e6108ba6192d0724d67f29deb0c9/tools/filesystem/filesystem.go#L70-L75

So I tried modifying it to always execute and it worked for me.

I'm wondering if it is possible to add an option for this workaround? or perhaps customizable ignored headers instead of a hard-coded endpoint check? Thanks.

**Comment by ganigeorgiev** [comment]
The above technically shouldn't be needed in the first place because it is not expected the provider to alter the headers.

`aws-sdk-go-v2` also ideally should work out of the box with all S3 compatible vendors but the reality unfortunately is different (_and it is the only PocketBase dependency that I wish I could remove because they often introduces breaking changes, slowdowns the initial build time and adds at least 10MB to the final binary; but for now this is put on hold_).

I'm fine with adding exceptions for other providers but I'm against making the headers option configurable as it is starting to become too complicated and messy.

I'm not familiar and not sure how to test SeaweedFS and I'm not planning working on it for now.

If you think that the current S3 handling can be improved feel free to open a PR in the PocketBase repo against the `develop` branch or ideally make an attempt to fix it in the upstream `aws-sdk-go-v2`.

**Reply by gera2ld** [reply]
Thanks for the quick reply.

> I'm not familiar and not sure how to test SeaweedFS and I'm not planning working on it for now.

It's a self-hosted service so there is no easy and reliable way to detect from Pocketbase.

An easy fix is to simply ignore `Accept-Encoding` for all backends (as in `aws-sdk-go`), which seems to be how [other projects](https://github.com/aws/aws-sdk-go-v2/issues/1816#ref-pullrequest-2091048479) fix this issue.
If you don't mind, I can open a PR to remove the `if` condition.

**Reply by ganigeorgiev** [reply]
> It's a self-hosted service so there is no easy and reliable way to detect from Pocketbase.

If it is a self-hosted service then you are in control and can try to prevent the headers change by finding out why it is being altered in the first place (e.g. reverse proxy misconfiguration or similar).

> An easy fix is to simply ignore Accept-Encoding for all backends...

No, because I don't know what other services this could break. If you think that this is an accepted solution then try to convince `aws-sdk-go-v2` to remove the header from the signature generation, which as you can see in the linked issue doesn't seem to have much interest because of:

> The V2 of the SDK was released to conform to requirements from the AWS service teams.

**Reply by gera2ld** [reply]
Finally confirmed `Accept-Encoding` is altered by Cloudflare from `identity` to `gzip, br`. :sweat_smile: 
~~I would compile a version with `Accept-Encoding` dropped as a workaround for now since I think this is the most sensible solution.~~

BTW I believe #5635 is the same issue as mine because the issue only occurs when using a custom domain with Cloudflare.

Thank you anyway for looking into the issue.

**Comment by gera2ld** [comment]
A summary for future visitors.

## Root Causes

1. **(RC1) Cloudflare alters `Accept-Encoding`**
   - `aws-sdk-go-v2` sets `Accept-Encoding: identity` and includes it in the signed headers.
   - Cloudflare overrides `Accept-Encoding: gzip, br`.
2. **(RC2) Cloudflare alters `HEAD` method into `GET` for caching purpose**
   - When deleting a file, the SDK will send a `HEAD` request first and get a `SignatureDoesNotMatch` error because the method doesn't match.

## Impacts

- `SignatureDoesNotMatch` behind Cloudflare and likely other reverse-proxies
- Other tools are impacted:
  - Supabase (https://github.com/supabase/storage/issues/577)

## Solutions

### (RC1) Override back `Accept-Encoding: identity` in your reverse proxy

**Requires a reverse proxy under control:**

> Pocketbase -> Cloudflare -> My Reverse Proxy -> My S3 Service

If I have a Caddy server between Cloudflare and my S3 service, I can do this:

```caddy
  reverse_proxy  {
    header_up Accept-Encoding identity
    to my-s3-service
  }
```

### (RC2) Disable cache in Cloudflare

Create a Cache rule in your dashboard, to bypass cache for all requests to your S3 server. `HEAD` requests will be preserved if cache is disabled.

## Other Solutions

### (RC1) `aws-sdk-go-v2` to exclude `Accept-Encoding` from signed headers

**Ideal, but unlikely.** (https://github.com/aws/aws-sdk-go-v2/issues/2848)

### (RC1) Patch Pocketbase to ignore `Accept-Encoding`

**Not ideal, but easy to achieve.**
